pkg/tsdb/postgres/postgres.go

```diff
@@ -61,15 +86,161 @@ func escape(input string) string {
+		if generatedFilePath == "" {
+			// Use standard filename
+			generatedFilePath = filepath.Join(dataDir, filename)
+		}
+
+		s.logger.Debug("Writing cert file", "path", generatedFilePath)
+		if err := ioutil.WriteFile(generatedFilePath, []byte(fileContent), 0600); err != nil {
+			return err
+		}
+		// Make sure the file has the permissions expected by the Postgresql driver, otherwise it will bail
+		if err := os.Chmod(generatedFilePath, 0600); err != nil {
+			return err
+		}
+		ds.JsonData.Set(jsonFieldName, generatedFilePath)
+		return nil
+	}
+
+	if generatedFilePath != "" {
+		s.logger.Debug("Deleting cert file since no content is provided", "path", generatedFilePath)
+		exists, err := fs.Exists(generatedFilePath)
+		if err != nil {
+			return err
+		}
+		if exists {
+			if err := os.Remove(generatedFilePath); err != nil {
+				return fmt.Errorf("failed to remove %q: %w", generatedFilePath, err)
+			}
+		}
+	}
+	ds.JsonData.Set(jsonFieldName, "")
+	return nil
+}
+
+func (s *postgresService) writeCertFiles(ds *models.DataSource) error {
+	s.logger.Debug("Writing TLS certificate files to disk")
+
+	decrypted := ds.SecureJsonData.Decrypt()
+	tlsRootCert := decrypted["tlsCACert"]
+	tlsClientCert := decrypted["tlsClientCert"]
+	tlsClientKey := decrypted["tlsClientKey"]
+
+	if tlsRootCert == "" && tlsClientCert == "" && tlsClientKey == "" {
+		s.logger.Debug("No TLS/SSL certificates provided")
+	}
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	workDir := filepath.Join(s.Cfg.DataPath, "tls", ds.Uid+"generatedTLSCerts")
+	exists, err := fs.Exists(workDir)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if err := os.MkdirAll(workDir, 0700); err != nil {
+			return err
+		}
+	}
+
+	if err := s.writeCertFile(ds, tlsRootCert, workDir, rootCert); err != nil {
+		return err
+	}
+	if err := s.writeCertFile(ds, tlsClientCert, workDir, clientCert); err != nil {
+		return err
+	}
+	if err := s.writeCertFile(ds, tlsClientKey, workDir, clientKey); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateCertFilePaths validates configured certificate file paths.
+var validateCertFilePaths = func(rootCert, clientCert, clientKey string) error {
+	for _, fpath := range []string{rootCert, clientCert, clientKey} {
+		if fpath == "" {
+			continue
+		}
+		exists, err := fs.Exists(fpath)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("certificate file %q doesn't exist", fpath)
+		}
+	}
+	return nil
+}
+
+func (s *postgresService) generateConnectionString(datasource *models.DataSource) (string, error) {
+	tlsConfigurationMethod := strings.TrimSpace(
+		strings.ToLower(datasource.JsonData.Get("tlsConfigurationMethod").MustString("file-path")))
 	tlsMode := strings.TrimSpace(strings.ToLower(datasource.JsonData.Get("sslmode").MustString("verify-full")))
 	isTLSDisabled := tlsMode == "disable"
 
+	if !isTLSDisabled && tlsConfigurationMethod == "file-content" {
+		if err := s.writeCertFiles(datasource); err != nil {
```

Comment:
Is there any reason writing the cert files and populating `datasource.JsonData` with generated cert files rather  than doing it at line 273 below?